src/runtime/tracing.cpp

```diff
@@ -183,16 +286,18 @@ WEAK trace_fn halide_set_custom_trace(trace_fn t) {
 
 WEAK void halide_set_trace_file(int fd) {
     halide_trace_file = fd;
-    halide_trace_file_initialized = true;
 }
 
 extern int errno;
 
 WEAK int halide_get_trace_file(void *user_context) {
-    // Prevent multiple threads both trying to initialize the trace
-    // file at the same time.
-    ScopedSpinLock lock(&halide_trace_file_lock);
-    if (!halide_trace_file_initialized) {
+    if (halide_trace_file >= 0) {
```

Comment:
This is double checked locking isn't it? Pretty sure need some sort of atomicity. One possibility is to use the same idea as the exclusive lock, but to store the fd in the low order bits. If it is a bad value, try to get it in exclusive mode, then the winner opens the file. (I'm presuming that contention for this spin lock is too expensive.

Reply:
I just moved the lock outwards. Didn't seem to hurt performance in any significant way.